Add new configure option --with-mixedrefs for mixed pointer mode

sharon-wang opened the pull request:
These are the jdk11 extensions changes needed for https://github.com/eclipse/openj9/issues/8878.

If `--with-mixedrefs` is specified in the configure command, the 'default' `OPENJ9_LIBS_SUBDIR` is used for the build. The mxdptrs CMake file corresponding to the build OS is used as the `OPENJ9_BUILDSPEC`. `--with-mixedrefs=[static/dynamic]` sets the reference determination mode. `--with-mixedrefs` without any value provided defaults to `static` mode.

The compressed and full GC libraries generated with the mixed build work are copied to the `OPENJ9_LIBS_SUBDIR` alongside the other existing libraries.
- `j9gc29`, `j9gcchk29` are always copied
- `j9gc_full29`, `j9gcchk_full29` are copied when running in mixedrefs `static` mode

Note that this will only work with an OpenJ9 CMake build. OpenJ9 UMA changes will be implemented after the CMake work is completed and any corresponding OMR/extensions changes will be provided in future PRs.

These changes depend on the new build specs defined in OpenJ9.

OpenJ9 PR: https://github.com/eclipse/openj9/pull/11166
OMR PR: https://github.com/eclipse/omr/pull/5657

Ports

- JDK8: https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/458
- JDK15: https://github.com/ibmruntimes/openj9-openjdk-jdk15/pull/43
- JDKNext: https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/250

sharon-wang commented:
Please review @gacholio

keithc-ca commented on closed/custom/copy/Copy-java.base.gmk:
```diff
@@ -59,11 +59,23 @@ $(call openj9_copy_files_and_debuginfos, \
 NO_JIT_CPUS := riscv64
 
+# Decide which GC lib(s) to copy
+POINTER_MODE := mixed
+ifeq (false,$(OPENJ9_ENABLE_MIXED_REFERENCES))
+	ifeq (compressedrefs,$(OPENJ9_LIBS_SUBDIR))
+		POINTER_MODE := compressed
+	else
+		POINTER_MODE := full
+	endif
+endif
```
Tabs must not precede macro definitions. Suggest expressing this as:
```
ifeq (true,$(OPENJ9_ENABLE_MIXED_REFERENCES))
  POINTER_MODE := mixed
else ifeq (compressedrefs,$(OPENJ9_LIBS_SUBDIR))
  POINTER_MODE := compressed
else
  POINTER_MODE := full
endif
```

keithc-ca commented on closed/autoconf/custom-hook.m4:
```diff
@@ -342,16 +342,26 @@ AC_DEFUN([OPENJ9_PLATFORM_SETUP],
-    OPENJ9_BUILDSPEC="${OPENJDK_BUILD_OS}_${OPENJ9_CPU}_cmprssptrs"
-    OPENJ9_LIBS_SUBDIR=compressedrefs
+  OPENJ9_ENABLE_MIXED_REFERENCES=false
+  if test "x$with_mixedrefs" = x; then
```
A user might say `--without-mixedrefs` which leads to `$with_mixedrefs = no`.

keithc-ca commented on closed/autoconf/custom-hook.m4:
```diff
@@ -342,16 +342,26 @@ AC_DEFUN([OPENJ9_PLATFORM_SETUP],
-    OPENJ9_LIBS_SUBDIR=compressedrefs
+  OPENJ9_ENABLE_MIXED_REFERENCES=false
+  if test "x$with_mixedrefs" = x; then
+    if test "x$with_noncompressedrefs" = x ; then
```
Same goes for `--without-noncompressedrefs`.

keithc-ca commented on closed/autoconf/custom-hook.m4:
```diff
@@ -342,16 +342,26 @@ AC_DEFUN([OPENJ9_PLATFORM_SETUP],
-    OPENJ9_BUILDSPEC="${OPENJDK_BUILD_OS}_${OPENJ9_CPU}_cmprssptrs"
-    OPENJ9_LIBS_SUBDIR=compressedrefs
+  OPENJ9_ENABLE_MIXED_REFERENCES=false
+  if test "x$with_mixedrefs" == xno -o "x$with_mixedrefs" == x; then
+    if test "x$with_noncompressedrefs" == xno -o "x$with_noncompressedrefs" == x ; then
```
Please use `=` instead of `==` for greater portability.

sharon-wang replied:
Sure thing. Is this also a good time to replace existing usage of `==` with `=` for the cmake and freemarker checks too?

keithc-ca replied:
Yes, that would be good to clean those up. Thanks.

keithc-ca commented on closed/autoconf/custom-hook.m4:
```diff
@@ -342,16 +342,26 @@ AC_DEFUN([OPENJ9_PLATFORM_SETUP],
     [build non-compressedrefs vm (large heap)])])
 
+  AC_ARG_WITH(mixedrefs, [AS_HELP_STRING([--with-mixedrefs],
+    [build mixedrefs vm ])])
```
nit: extra space after 'vm'

keithc-ca commented on closed/autoconf/custom-hook.m4:
```diff
@@ -342,16 +342,26 @@ AC_DEFUN([OPENJ9_PLATFORM_SETUP],
-    OPENJ9_BUILDSPEC="${OPENJDK_BUILD_OS}_${OPENJ9_CPU}_cmprssptrs"
-    OPENJ9_LIBS_SUBDIR=compressedrefs
+  OPENJ9_ENABLE_MIXED_REFERENCES=false
+  if test "x$with_mixedrefs" = xno -o "x$with_mixedrefs" = x; then
+    if test "x$with_noncompressedrefs" = xno -o "x$with_noncompressedrefs" = x ; then
+      OPENJ9_BUILDSPEC="${OPENJDK_BUILD_OS}_${OPENJ9_CPU}_cmprssptrs"
+      OPENJ9_LIBS_SUBDIR=compressedrefs
+    else
+      OPENJ9_BUILDSPEC="${OPENJDK_BUILD_OS}_${OPENJ9_CPU}"
+      OPENJ9_LIBS_SUBDIR=default
+    fi
   else
-    OPENJ9_BUILDSPEC="${OPENJDK_BUILD_OS}_${OPENJ9_CPU}"
+    OPENJ9_ENABLE_MIXED_REFERENCES=true
+    OPENJ9_BUILDSPEC="${OPENJDK_BUILD_OS}_${OPENJ9_CPU}_mxdptrs"
     OPENJ9_LIBS_SUBDIR=default
   fi
```
I think this reads better if you use `if` and `elif`:
```
  if test "x$with_mixedrefs" = xyes ; then
    OPENJ9_BUILDSPEC="${OPENJDK_BUILD_OS}_${OPENJ9_CPU}_mxdptrs"
    OPENJ9_ENABLE_MIXED_REFERENCES=true
    OPENJ9_LIBS_SUBDIR=default
  elif test "x$with_noncompressedrefs" = xyes ; then
    OPENJ9_BUILDSPEC="${OPENJDK_BUILD_OS}_${OPENJ9_CPU}_cmprssptrs"
    OPENJ9_ENABLE_MIXED_REFERENCES=false
    OPENJ9_LIBS_SUBDIR=compressedrefs
  else
    OPENJ9_BUILDSPEC="${OPENJDK_BUILD_OS}_${OPENJ9_CPU}"
    OPENJ9_ENABLE_MIXED_REFERENCES=false
    OPENJ9_LIBS_SUBDIR=default
  fi
```
(one level of nesting, setting variables in the same order)

keithc-ca commented on closed/autoconf/custom-hook.m4:
```diff
@@ -360,25 +370,37 @@ AC_DEFUN([OPENJ9_PLATFORM_SETUP],
       OPENJ9_PLATFORM_CODE=wa64
-      if test "x$OPENJ9_LIBS_SUBDIR" = xdefault ; then
-        OPENJ9_BUILDSPEC=win_x86-64
+      if test "x$OPENJ9_ENABLE_MIXED_REFERENCES" = xfalse ; then
+        if test "x$OPENJ9_LIBS_SUBDIR" = xdefault ; then
+          OPENJ9_BUILDSPEC=win_x86-64
+        else
+          OPENJ9_BUILDSPEC=win_x86-64_cmprssptrs
+        fi
```
Rather than repeating tests for the three modes, all we really need to do is change the prefix of `OPENJ9_BUILDSPEC` from 'windows' to 'win'; likewise for macosx (removing 'mac').

keithc-ca commented:
> Note that this will only work with an OpenJ9 CMake

The configure script should complain (and fail) if you try to combine `--without-cmake` and `--with-mixedrefs`.

sharon-wang commented:
Thanks @keithc-ca!

Added a check so that `configure` errors and quits if mixedrefs is used while CMake is not enabled. It's in the CMake configure function since it runs after the platform setup, so at that point we'll know if mixedrefs is set and whether CMake is enabled or not.
https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/b944640ba16cdd78955ecec546be4512c990df63/closed/autoconf/custom-hook.m4#L75-L82

For setting `OPENJ9_BUILDSPEC`, I've split the build spec into the OS portion `OPENJ9_BUILD_OS`/`OPENJDK_BUILD_OS` and the mode/arch portion `OPENJ9_BUILD_MODE_ARCH`. I concatenate the build OS and the build mode/arch after doing all the tests to construct `OPENJ9_BUILDSPEC`. Let me know what you think of this approach.

keithc-ca commented on closed/autoconf/custom-hook.m4:
```diff
@@ -360,52 +374,56 @@ AC_DEFUN([OPENJ9_PLATFORM_SETUP],
     AC_MSG_ERROR([Unsupported OpenJ9 cpu ${OPENJ9_CPU}!])
   fi
 
+  if test "x$OPENJ9_BUILD_OS" != x ; then
+    OPENJ9_BUILDSPEC="${OPENJ9_BUILD_OS}_${OPENJ9_BUILD_MODE_ARCH}"
+  else
+    OPENJ9_BUILDSPEC="${OPENJDK_BUILD_OS}_${OPENJ9_BUILD_MODE_ARCH}"
+  fi
```
If you assign the default above (`OPENJ9_BUILD_OS=$OPENJDK_BUILD_OS`) then this can be unconditional:
```
  OPENJ9_BUILDSPEC="${OPENJ9_BUILD_OS}_${OPENJ9_BUILD_MODE_ARCH}"
```

keithc-ca commented on closed/autoconf/custom-hook.m4:
```diff
@@ -75,6 +75,11 @@ AC_DEFUN([OPENJ9_CONFIGURE_CMAKE],
+
+    # Currently, mixedrefs mode is only available with CMake enabled
+    if test "x$OPENJ9_ENABLE_MIXED_REFERENCES" = xtrue ; then
+      AC_MSG_ERROR([--with-mixedrefs cannot be used if CMake is disabled])
```
Can you make this message instructive, perhaps, '--with-mixedrefs requires --with-cmake`?

keithc-ca commented:
I think this is ready to merge: please squash.

Will you be preparing pull requests for the other extension repositories?

> will be opening a PR with these changes soon

When should we expect that?

sharon-wang commented:
I think the focus is on JDK11 right now, but let me confirm in the main issue. I think this depends on the availability of CMake in JDK versions/various platforms.

For the openj9 PR, I'm working on finishing up the pipeline changes needed (and testing them) and reorganizing the commits since there's a lot of changes. I expect to open up a PR by mid next week.

I've opened up a PR https://github.com/eclipse/omr/pull/5657 for the OMR GC changes.

sharon-wang commented:
I've squashed the commits.

Are there documentation changes that are needed with this new configure option in the extensions repo? Or just within openj9/openj9-docs?

keithc-ca commented:
This new option would be documented in the [build instructions for Java 11](https://github.com/eclipse/openj9/blob/master/doc/build-instructions/Build_Instructions_V11.md), and similarly for other versions, probably near places where `--with-noncompressedrefs` is mentioned.

gacholio commented:
Would it be worth splitting the mixed refs and dual GC compile options?

keithc-ca commented:
> Would it be worth splitting the mixed refs and dual GC compile options?

I don't understand what that means. As I understand it, mixed refs implies dual GC compiles. What could be split?

gacholio commented:
Mixed refs work fine with the single compile of the GC, it just doesn't perform well. In the future, we may want to invest some time in templating, etc. to allow the single compile to work efficiently.

sharon-wang commented:
For splitting, are you referring to adding pointer mode checks?

keithc-ca commented:
If supporting both flavours of mixed refs isn't too onerous, it probably makes sense to do that now. I suggest we expose that option via using _value_ of `--with-mixedrefs=value` to identify the desired mode:
* blank or 'no' - mixed references are not enabled
* 'yes' or 'static' - mixed references are enabled using dual compiles
* 'dynamic' - mixed references are enabled with dynamic (runtime) tests

The values 'static' and 'dynamic' are just my suggestions: others may have better suggestions.

sharon-wang commented:
Ah I see. So in 'dynamic' mode, we actually don't want to define `OMR_OVERRIDE_COMPRESS_OBJECT_REFERENCES`.

I can set a new CMAKE argument `-DMIXEDREFS_MODE=[static/dynamic]`, which is read by the CMAKE files in omr to conditionally set the OVERRIDE value.

gacholio commented:
Correct, dynamic mode just defines both the compressed and full defines, without the override.

keithc-ca commented on closed/OpenJ9.gmk:
```diff
@@ -341,6 +341,7 @@ ifeq (true,$(OPENJ9_ENABLE_CMAKE))
 	-DOPENJ9_BUILD=true \
 	-DOPENJ9_SHA=$(OPENJ9_SHA) \
 	-DOPENJDK_VERSION_NUMBER_FOUR_POSITIONS=$(VERSION_NUMBER_FOUR_POSITIONS) \
+	-DOMR_MIXED_REFERENCES_MODE="$(OMR_MIXED_REFERENCES_MODE)"\
```
nit: space before `\`

keithc-ca commented on closed/autoconf/custom-spec.gmk.in:
```diff
@@ -85,6 +85,10 @@ OPENJ9_PLATFORM_CODE    := @OPENJ9_PLATFORM_CODE@
 OPENJ9_LIBS_SUBDIR      := @OPENJ9_LIBS_SUBDIR@
 
+# Mixed References Mode
+OPENJ9_ENABLE_MIXED_REFERENCES := @OPENJ9_ENABLE_MIXED_REFERENCES@
+OMR_MIXED_REFERENCES_MODE   := @OMR_MIXED_REFERENCES_MODE@
```
I don't think we need both of these, if `OMR_MIXED_REFERENCES_MODE` is one of `off` (or something similar), `static` or `dynamic`.

sharon-wang commented:
I've added `OMR_MIXED_REFERENCES_MODE` which is off by default, but can be set to `static` or `dynamic` if `--with-mixedrefs=[static/dynamic]` is used. `--with-mixedrefs` defaults to `static`. This will be documented in OpenJ9 `doc/build-instructions/Build_Instructions_V[8/11/15/...].md`

If we are in mixed mode and `static` is the setting, then we will set `OMR_OVERRIDE_COMPRESS_OBJECT_REFERENCES` in OMR.

If these changes look good, I'll port them to the JDK8/15/next equivalents.

keithc-ca commented on closed/autoconf/custom-hook.m4:
```diff
@@ -75,6 +75,11 @@ AC_DEFUN([OPENJ9_CONFIGURE_CMAKE],
+
+    # Currently, mixedrefs mode is only available with CMake enabled
+    if test "x$OMR_MIXED_REFERENCES_MODE" != xoff ; then
+      AC_MSG_ERROR([--with-mixedrefs=[static|dynamic] requires --with-cmake])
```
An extra set of square brackets are needed to retain the ones in `[static|dynamic]`:
```
  AC_MSG_ERROR([[--with-mixedrefs=[static|dynamic] requires --with-cmake]])
```

keithc-ca commented on closed/OpenJ9.gmk:
```diff
@@ -341,6 +341,7 @@ ifeq (true,$(OPENJ9_ENABLE_CMAKE))
 	-DOPENJ9_BUILD=true \
 	-DOPENJ9_SHA=$(OPENJ9_SHA) \
 	-DOPENJDK_VERSION_NUMBER_FOUR_POSITIONS=$(VERSION_NUMBER_FOUR_POSITIONS) \
+	-DOMR_MIXED_REFERENCES_MODE="$(OMR_MIXED_REFERENCES_MODE)" \
```
Please move this up to retain sorted order of variable definitions (after `OMR_DDR`).
I don't think the quotes are necessary (the value will be one of `off`, `static` or `dynamic`).

keithc-ca commented on closed/custom/copy/Copy-java.base.gmk:
```diff
@@ -59,11 +59,22 @@ $(call openj9_copy_files_and_debuginfos, \
 # CPU targets without JIT support.
 NO_JIT_CPUS := riscv64
 
+# Decide which GC lib(s) to copy
+ifneq (off,$(OMR_MIXED_REFERENCES_MODE))
+  POINTER_MODE := mixed
+else ifeq (compressedrefs,$(OPENJ9_LIBS_SUBDIR))
+  POINTER_MODE := compressed
+else
+  POINTER_MODE := full
+endif
```
nit: please end a sentence with a period.

I don't think this handles `OMR_MIXED_REFERENCES_MODE=dynamic` properly. In that case I expect only one GC library (`j9gc29`), but the code below will try to copy the 'full' variants as well.

sharon-wang replied:
Ah, my understanding was a bit off. Right, in dynamic mode there's no point in generating two libraries because there's no OVERRIDE to set, so there will be no difference in the builds. Will be updating this and fixing it in my OMR/OpenJ9 changes too

sharon-wang commented:
I've added the requested changes and updated the description to note:
- `j9gc29`, `j9gcchk29` are always copied
- `j9gc_full29`, `j9gcchk_full29` are copied when running in mixedrefs static mode

keithc-ca reviewed (changes requested):
Not as part of this pull request, but at some point I think we should have just one configuration option that affects reference representation (not both `--with-mixedrefs` and `--with-noncompressedrefs`), perhaps `--with-references=[full|compressed-static|compressed-dynamic]` so we don't have to consider precedence of the two options.

sharon-wang commented:
Updated. Help string is now `build mixedrefs vm (--with-mixedrefs=static or --with-mixedrefs=dynamic)`

For --with-references=[full|**compressed-static|compressed-dynamic**], do you mean `mixed-static|mixed-dynamic`?

gacholio commented:
> Updated. Help string is now `build mixedrefs vm (--with-mixedrefs=static or --with-mixedrefs=dynamic)`
> 
> For --with-references=[full|**compressed-static|compressed-dynamic**], do you mean `mixed-static|mixed-dynamic`?

We'll need `[full|compressed|mixed-static|mixed-dynamic]`

sharon-wang commented:
I opened https://github.com/ibmruntimes/openj9-openjdk-jdk11/issues/361 to keep note of this new configure option.

keithc-ca commented:
What changes, required by this, are yet to be merged into openj9 and/or omr?

gacholio reviewed (approved):
I'll defer to Keith here.

sharon-wang commented:
@keithc-ca Here are the corresponding PRs
- **OpenJ9:** https://github.com/eclipse/openj9/pull/11166
- **OMR:** https://github.com/eclipse/omr/pull/5657

keithc-ca commented:
Jenkins compile aix,osx,win,plinux jdk11
